Mathlib/Topology/UniformSpace/DiscreteUniformity.lean

```diff
@@ -0,0 +1,102 @@
+/-
+Copyright (c) 2024 Antoine Chambert-Loir, María Inés de Frutos Fernández. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johannes Hölzl, Mario Carneiro, Antoine Chambert-Loir, María Inés de Frutos Fernández
+-/
+import Mathlib.Topology.UniformSpace.Cauchy
+import Mathlib.Topology.Algebra.UniformGroup.Defs
+
+/-! # Discrete uniformity
+
+The discrete uniformity is the smallest possibly uniformity, the one for which
+the diagonal is an entourage of itself.
+
+It induces the discrete topology.
+
+It is complete.
+
+-/
+
+open Filter UniformSpace
+
+/-- The discrete uniformity -/
+class DiscreteUniformity (X : Type*) [u : UniformSpace X] : Prop where
+  eq_bot: u = ⊥
+
+namespace DiscreteUniformity
+
+/-- The bot uniformity is the discrete uniformity -/
+instance (X : Type*) : @DiscreteUniformity X ⊥ :=
+  @DiscreteUniformity.mk X ⊥ rfl
+
+variable (X : Type*) [u : UniformSpace X] [DiscreteUniformity X]
+
+theorem eq_principal_idRel : uniformity X = principal idRel :=
+  le_antisymm
+    (by simpa [uniformSpace_eq_bot, ← le_principal_iff] using eq_bot)
+    refl_le_uniformity
+
+theorem iff_eq_principal_idRel {X : Type*} [UniformSpace X] :
+    DiscreteUniformity X ↔ uniformity X = principal idRel :=
+  ⟨fun _ ↦ eq_principal_idRel X, fun h ↦ { eq_bot := by ext; rw [h]; rfl }⟩
+
+/-- The discrete uniformity induces the discrete topology -/
+instance : DiscreteTopology X where
+  eq_bot := by
+    convert UniformSpace.toTopologicalSpace_bot
+    exact eq_bot
+
+theorem idRel_mem_uniformity : idRel ∈ uniformity X := by
+  simp only [eq_principal_idRel, mem_principal, subset_refl]
+
+variable {X} in
+/-- A product of spaces with discrete uniformity has a discrete uniformity -/
+instance {Y : Type*} [UniformSpace Y] [DiscreteUniformity Y] :
+    DiscreteUniformity (X × Y) where
+  eq_bot := by
+    rw [instUniformSpaceProd, UniformSpace.uniformSpace_eq_bot, uniformity_prod_eq_comap_prod,
+      mem_comap]
+    refine ⟨idRel.prod idRel, prod_mem_prod (idRel_mem_uniformity _) (idRel_mem_uniformity _), ?_⟩
+    rintro ⟨⟨x, y⟩, x', y'⟩ h
+    simpa only [mem_idRel, Prod.mk.injEq] using h
+
+variable {x} in
+/-- On a space with a discrete uniformity, any function is uniformly continuous -/
+theorem uniformContinuous {Y : Type*} [UniformSpace Y] (f : X → Y) :
+    UniformContinuous f := fun s hs ↦ by
+  rw [mem_map]
+  apply Filter.mem_of_superset (idRel_mem_uniformity X)
+  simp only [idRel_subset, Set.mem_preimage]
+  exact fun _ ↦ mem_uniformity_of_eq hs rfl
```

Comment:
I can't find it right now, but there should be a lemma saying that uniform continuity is the same as some inequality of uniform structures, could you use that?

Reply:
I think @ADedecker hoped for that
```
variable {x} in
/-- On a space with a discrete uniformity, any function is uniformly continuous -/
theorem uniformContinuous {Y : Type*} [UniformSpace Y] (f : X → Y) :
    UniformContinuous f := by
  simp only [uniformContinuous_iff, DiscreteUniformity.eq_bot, bot_le]

```

Reply:
Indeed, Antoine's version is what I had in mind.